### ISSUE TITLE
latest Nvidia drivers cause system freeze on exit of Chromium or Electron-based programs

### ISSUE THREAD

**Issue by seankhl** [void-linux/void-packages#32241]
### System
* xuname: ``Void 5.12.19_1 x86_64 AuthenticAMD uptodate rrrrmmnDFFFFFFFFFFFFFF``
* *affected package(s) including the version*: `nvidia-470.57.02_1` (etc., the set of 470.57.02_1 nvidia packages)

I'm using a 1050 Ti.

### Expected behavior
No crash

### Actual behavior
Upon upgrading from the 460.84_2 Nvidia drivers, I am having an issue with any Electron-based program (including Chromium) where the system will freeze when I close the program. Sometimes the audio will still play, but my keyboard and mouse do nothing and the audio eventually freezes too and I have to power cycle my machine. I'm not seeing anything in the logs so I'm not sure how to diagnose the problem. Happy to get any advice to figure out how to find some warning, error, etc. I also tried to `xdowngrade` the following packages which are still in my cache:

```bash
# xdowngrade \
        /var/cache/xbps/nvidia-460.84_2.x86_64.xbps \
        /var/cache/xbps/nvidia-dkms-460.84_2.x86_64.xbps \
        /var/cache/xbps/nvidia-gtklibs-460.84_2.x86_64.xbps \
        /var/cache/xbps/nvidia-libs-460.84_2.x86_64.xbps \
        /var/cache/xbps/nvidia-libs-32bit-460.84_2.x86_64.xbps \
        /var/cache/xbps/nvidia-opencl-460.84_2.x86_64.xbps
```

but I must have done something wrong, the graphical login failed and I had to boot to a root shell to re-update my drivers to get a working system.

If I `kill` or `pkill` the program it will close without any problems (hard to remember to do, leading to a lot system freezes).

### Steps to reproduce the behavior
For me, it was as simple as upgrading the Nvidia drivers and rebooting.

**Comment by seanlaguna** [comment]
I was able to downgrade by issuing the above `xdowngrade` command and then copying the kernel modules manually into the appropriate directory:
```bash
# cp /var/lib/dkms/nvidia/460.84/5.12.19_1/x86_64/module/* /lib/modules/5.12.19_1/kernel/drivers/video/
```

Before that Xorg crashed with an API mismatch on the Nvidia kernel modules (still at 470.57.02). I'm not sure why the drivers weren't copied automatically when the downgrade happened.

**Comment by ericonr** [comment]
DKMS and downgrading don't mix very well, unfortunately.

What logs have you checked? `dmesg` should contain something, at least. Does your setup use PRIME or not?

@abenson

**Comment by Lolzen** [comment]
i do have the same issue.
may be related to #32222 in some way

**Comment by abenson** [comment]
I can replicate this with `prime-run slack`.  According to the [nvidia forums](https://forums.developer.nvidia.com/t/google-chrome-92-segfaults-in-nvidia-drivers-470-57-02/184244/3), this is being tracked and already fixed.

**Comment by reedts** [comment]
Same problem on my system on `Signal-Desktop` or `element-desktop`. AMD Ryzen 7 3700X, Nvidia RTX 3060Ti.

output of dmesg after freeze:
```
[ 9092.635328] general protection fault, probably for non-canonical address 0x6b6b6b6b00000008: 0000 [#1] SMP NOPTI
[ 9092.635334] CPU: 13 PID: 1627 Comm: signal-desktop Tainted: P           O      5.13.6_1 #1
[ 9092.635336] Hardware name: Micro-Star International Co., Ltd. MS-7C91/MAG B550 TOMAHAWK (MS-7C91), BIOS A.30 09/03/2020
[ 9092.635337] RIP: 0010:_nv035844rm+0xb0/0xe0 [nvidia]
[ 9092.635586] Code: 89 c2 48 89 ef 48 8d b1 48 01 00 00 4c 89 e9 e8 a6 5b ff ff 66 0f 1f 44 00 00 48 89 ef e8 08 5c ff ff 84 c0 74 8a 48 8b 75 00 <48> 39 5e 08 75 ea 4c 39 26 75 e5 49 8b 44 24 20 48 8d b8 48 01 00
[ 9092.635588] RSP: 0018:ffffac1f43a37bc8 EFLAGS: 00010202
[ 9092.635590] RAX: 0000000000000001 RBX: ffff9320cadf8830 RCX: ffff9320b0ed9178
[ 9092.635591] RDX: 6b6b6b6b6b6b6b6b RSI: 6b6b6b6b00000000 RDI: ffff932088c22d20
[ 9092.635592] RBP: ffff932088c22d20 R08: 0000000000000020 R09: ffff932088c22d28
[ 9092.635593] R10: 0000000000000000 R11: 0000000000000001 R12: ffff9320cacd4cb0
[ 9092.635593] R13: 6b6b6b6b00000000 R14: ffff932088c22d98 R15: ffff9320cadf8830
[ 9092.635595] FS:  0000000000000000(0000) GS:ffff93278ed40000(0000) knlGS:0000000000000000
[ 9092.635596] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 9092.635597] CR2: 00007fd424ba4ad8 CR3: 000000039e610000 CR4: 0000000000350ee0
[ 9092.635598] Call Trace:
[ 9092.635601]  ? _nv014655rm+0x2ee/0x770 [nvidia]
[ 9092.635800]  ? _nv037695rm+0xb3/0x150 [nvidia]
[ 9092.635997]  ? _nv037694rm+0x297/0x4e0 [nvidia]
[ 9092.636195]  ? _nv037689rm+0x60/0x70 [nvidia]
[ 9092.636392]  ? _nv037690rm+0x7b/0xb0 [nvidia]
[ 9092.636590]  ? _nv036056rm+0x40/0xe0 [nvidia]
[ 9092.636779]  ? _nv000699rm+0x68/0x80 [nvidia]
[ 9092.636954]  ? rm_cleanup_file_private+0xea/0x160 [nvidia]
[ 9092.637127]  ? nvidia_close+0x150/0x310 [nvidia]
[ 9092.637295]  ? nvidia_frontend_close+0x2b/0x50 [nvidia]
[ 9092.637465]  ? __fput+0x8c/0x230
[ 9092.637468]  ? task_work_run+0x62/0x90
[ 9092.637471]  ? do_exit+0x367/0xab0
[ 9092.637473]  ? do_group_exit+0x33/0xa0
[ 9092.637475]  ? __x64_sys_exit_group+0x14/0x20
[ 9092.637476]  ? do_syscall_64+0x61/0x80
[ 9092.637479]  ? do_syscall_64+0x6e/0x80
[ 9092.637480]  ? asm_exc_page_fault+0x8/0x30
[ 9092.637481]  ? entry_SYSCALL_64_after_hwframe+0x44/0xae
[ 9092.637483] Modules linked in: snd_seq_dummy snd_hrtimer 8021q cfg80211 garp mrp stp llc nls_iso8859_1 nls_cp437 vfat fat nvidia_drm(PO) nvidia_modeset(PO) nvidia(PO) intel_rapl_msr intel_rapl_common snd_hda_codec_realtek snd_hda_codec_generic edac_mce_amd ledtrig_audio snd_hda_codec_hdmi drm_kms_helper snd_hda_intel kvm_amd snd_intel_dspcfg cec snd_intel_sdw_acpi rc_core snd_hda_codec kvm irqbypass drm input_leds joydev rapl snd_hda_core pcspkr snd_hwdep wmi_bmof r8169 agpgart syscopyarea k10temp snd_pcm sysfillrect realtek sysimgblt fb_sys_fops sp5100_tco i2c_piix4 evdev gpio_amdpt acpi_cpufreq mac_hid gpio_generic tiny_power_button snd_seq snd_seq_device snd_timer snd soundcore vhost_vsock vmw_vsock_virtio_transport_common vsock vhost_net vhost vhost_iotlb tap uhid hci_vhci bluetooth ecdh_generic rfkill ecc vfio_iommu_type1 vfio uinput userio ppp_generic slhc tun loop nvram cuse fuse ext4 crc16 mbcache jbd2 usbkbd hid_generic usbmouse usbhid hid ahci xhci_pci xhci_pci_renesas
[ 9092.637520]  crct10dif_pclmul libahci crc32_pclmul xhci_hcd ghash_clmulni_intel aesni_intel libata crypto_simd cryptd ccp usbcore scsi_mod rng_core wmi button dm_mirror dm_region_hash dm_log dm_mod btrfs blake2b_generic xor raid6_pq libcrc32c crc32c_generic crc32c_intel
[ 9092.637551] ---[ end trace 5c591019667d32f9 ]---
[ 9092.637552] RIP: 0010:_nv035844rm+0xb0/0xe0 [nvidia]
[ 9092.637811] Code: 89 c2 48 89 ef 48 8d b1 48 01 00 00 4c 89 e9 e8 a6 5b ff ff 66 0f 1f 44 00 00 48 89 ef e8 08 5c ff ff 84 c0 74 8a 48 8b 75 00 <48> 39 5e 08 75 ea 4c 39 26 75 e5 49 8b 44 24 20 48 8d b8 48 01 00
[ 9092.637813] RSP: 0018:ffffac1f43a37bc8 EFLAGS: 00010202
[ 9092.637815] RAX: 0000000000000001 RBX: ffff9320cadf8830 RCX: ffff9320b0ed9178
[ 9092.637816] RDX: 6b6b6b6b6b6b6b6b RSI: 6b6b6b6b00000000 RDI: ffff932088c22d20
[ 9092.637817] RBP: fff932088c22d20 R08: 0000000000000020 R09: ffff932088c22d28
[ 9092.637818] R10: 0000000000000000 R11: 0000000000000001 R12: ffff9320cacd4cb0
[ 9092.637820] R13: 6b6b6b6b00000000 R14: ffff932088c22d98 R15: ffff9320cadf8830
[ 9092.637821] FS:  0000000000000000(0000) GS:ffff93278ed40000(0000) knlGS:0000000000000000
[ 9092.637822] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 9092.637824] CR2: 00007fd424ba4ad8 CR3: 000000039e610000 CR4: 0000000000350ee0
[ 9092.637826] Fixing recursive fault but reboot is needed!
```

**Comment by mvf** [comment]
The `0x6b` byte in the fault address and a bunch of regs is the kernel's `POISON_FREE` value, a tell-tale sign of use-after-free. In the Void kernel, this poisoning is only performed when `slub_debug=P` is set on the kernel command line. It has measurable performance overhead, that's why nobody uses it in production. It was recently removed from the defaults (155b23d17543c31b5617bd6ea1943864370650ad), but stays enabled on existing installations.

Removing `slub_debug=P` from the kernel command line will most likely make the Oops disappear. Of course that's undefined behavior, but OTOH, that's how other distros are running this driver.

**Comment by reedts** [comment]
@mvf Thank you very much! Solved the problem for me (also learned something!)

**Comment by seankhl** [comment]
Anyone knows the status of this with respect to 470.63.01_1?

**Comment by mvf** [comment]
> Anyone knows the status of this with respect to 470.63.01_1?

The issue persists.

**Comment by RononDex** [comment]
Having the same issue

**Comment by annada-behera** [comment]
One workaround is to disable GPU rendering for Chromium, Electron and it's children like Telegram and Signal, with the `--disable-gpu` flag. The system will not freeze nomore but you will not be able to use GPU for your browser.

**Comment by sloosch** [comment]
Somehow electron based apps seems to suffer from a use-after-free error when using hardware acceleration with nvidia drivers. Your system was maybe booted with some security measure to "poison" freed memory which could cause your system to hang.

Those flags were recently removed https://github.com/void-linux/void-packages/commit/155b23d17543c31b5617bd6ea1943864370650ad but maybe still active on your system.

In the defaults of grub (/etc/default/grub) alter the GRUB_CMDLINE_LINUX_DEFAULT to not contain "slub_debug" and "page_poison" option. Then recreate the grub.cfg via "sudo grub-mkconfig -o /boot/grub/grub.cfg"

**Comment by sloosch** [comment]
Had the same problem, can confirm @mvf what said:
In the defaults of grub (/etc/default/grub) alter the GRUB_CMDLINE_LINUX_DEFAULT to not contain "slub_debug" and "page_poison" option. Then recreate the grub.cfg via "sudo grub-mkconfig -o /boot/grub/grub.cfg"

**Comment by vikigenius** [comment]
Just removing the slub_debug option worked for me. page_poison doesn't seem to have any effect.

**Comment by abenson** [comment]
If the issue persists for you all, can you build and test [this pr, #33041](https://github.com/void-linux/void-packages/pull/33041)

**Comment by RononDex** [comment]
For me the issue still persists

**Comment by mvf** [comment]
Issue persists in local build of nvidia-495.44 on kernel 5.14.15.

**Comment by Duncaen** [comment]
Is there a point in keeping this open or continuing the process even though the kernel command line parameter is not enabled by default anymore?

**Comment by ericonr** [comment]
A bug report to upstream might help to make them aware the issue exists at all...

**Comment by kiwec** [comment]
> Is there a point in keeping this open or continuing comment on the progess even though the kernel command line parameter is not enabled by default anymore?

It's a use-after-free. Nvidia should fix their shit. Leaving that bug in "because it doesn't always crash" is insane.

**Comment by Duncaen** [comment]
> It's a use-after-free. Nvidia should fix their shit. Leaving that bug in "because it doesn't always crash" is insane.

Do we look like nvidia employees? There is nothing we can do about it, there is no point in this issue being open.

**Comment by kiwec** [comment]
> > It's a use-after-free. Nvidia should fix their shit. Leaving that bug in "because it doesn't always crash" is insane.
> 
> Do we look like nvidia employees? There is nothing we can do about it, there is no point in this issue being open.

This is not the first nvidia driver bug (and certainly not the last). It's better to keep the issue open so existing users know why their system is crashing, even if we have to wait for a fix for months (as has happened before).

Some people might feel fine leaving a potential vulnerability in their system, others will not.

**Comment by Duncaen** [comment]
Why should we provide issue trackers for proprietary software, someone should report it to nvidia and use that to track the status. This has nothing to do with void linux.

**Comment by dm17** [comment]
> Why should we provide issue trackers for proprietary software, someone should report it to nvidia and use that to track the status. This has nothing to do with void linux.

I don't use any nvidia stuff in Linux, but isn't it good for Void to track which software offends Void the most? That way Void will sort of promote the better software naturally? Especially considering Void's overall approach of changing software the least vs other distros.

**Comment by seankhl** [comment]
Anyone know if this has improved with the latest drivers? `xorg-server` now conflicts with `nvidia-450.119.03_1` (the last drivers without this issue).

**Comment by vanessa-vd98** [comment]
I have `nvidia-510.47.03_1` and my X doesn't crash when I close Chromium/Electron.

> Anyone know if this has improved with the latest drivers? `xorg-server` now conflicts with `nvidia-450.119.03_1` (the last drivers without this issue).

**Comment by RononDex** [comment]
The mentoined fix by @sloosch worked for me
> Had the same problem, can confirm what @mvf said: In the defaults of grub (/etc/default/grub) alter the GRUB_CMDLINE_LINUX_DEFAULT to not contain "slub_debug" and "page_poison" option. Then recreate the grub.cfg via "sudo grub-mkconfig -o /boot/grub/grub.cfg"

**Comment by kiwec** [comment]
> Anyone know if this has improved with the latest drivers? `xorg-server` now conflicts with `nvidia-450.119.03_1` (the last drivers without this issue).

Nope, still not fixed. Workaround works, but is not ideal.

**Comment by seankhl** [comment]
> Nope, still not fixed. Workaround works, but is not ideal.

Weird. I upgraded last night and the issue actually has resolved for me: closing chromium, discord, slack, etc. don't trigger a freeze. I only have `slub_debug=P` in my boot params though, not `page_poison`.

Fully agree that the workaround is not ideal.

**Comment by kiwec** [comment]
Fixed in `nvidia-515.65.01_1`!
